shell/platform/android/io/flutter/view/FlutterMain.java

```diff
@@ -178,31 +179,28 @@ public static void ensureInitializationComplete(Context applicationContext, Stri
             return;
         }
         try {
-            sResourceExtractor.waitForCompletion();
+            if (sResourceExtractor != null) {
+                sResourceExtractor.waitForCompletion();
```

Comment:
What does it mean in this situation and the one that follows if `sResourceExtractor` is null?

Reply:
@blasten it looks like this question remains. I asked because I'm not sure `sResourceExtractor` can be null, or if it can then it probably means we have a problem, right?

Reply:
well, it's `null` when there's isn't any resource to extract. (Added a comment). I'm not sure if we still want to construct the object. But I can see an argument for just constructing the object.

Reply:
What I'm curious about is what it means for us to be in these methods and no `sResourceExtractor` exists. Should it exist? If so, why doesn't it exist?

I'm guessing that silently bypassing the `waitForCompletion()` is probably not the right answer, but you can let me know if I'm missing something.

Reply:
Maybe to avoid the work when it's not needed. But if `start()` isn't called, then `waitForCompletion()` is a noop. https://github.com/flutter/engine/blob/0b69c54cf5254f2131f5d6511816042d4b05233b/shell/platform/android/io/flutter/view/ResourceExtractor.java#L183

Reply:
So I read through the code to see where `sResourceExtractor` is instantiated vs where it is used. I see that before this PR it is impossible for it to be null, but now it might be null because of the specific change being made in this PR. I think if you'd like to follow up later with some rework then all that is necessary here are some comments on those if-statements as to why it might be null. 

But I would also recommend filing an issue to come back and actually separate these execution paths. Rather than invoke the same functions that do slightly different things based on the packaging scheme, I would recommend 2 different objects that are responsible for bootstrapping Dart: one for native libraries, and one for snapshots. Then you can simply delegate at the beginning to one or the other and the reader of the code doesn't have to be aware of both paths in every function that follows. Does that make sense?

Reply:
Comment added, and issue filed: https://github.com/flutter/flutter/issues/33054